File: src/test/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserializerTest.scala
package com.exasol.cloudetl.kafka.deserialization

import java.nio.charset.StandardCharsets

import com.exasol.cloudetl.kafka.KafkaConnectorException
import com.exasol.common.json.JsonMapper

import com.fasterxml.jackson.databind.JsonNode
import org.apache.kafka.common.serialization.StringDeserializer
import org.scalatest.funsuite.AnyFunSuite

class JsonDeserializerTest extends AnyFunSuite {

  test("must deserialize json record with primitives") {
    val row = new JsonDeserializer(
      Seq(new RecordValueField("number"), new RecordValueField("string"), new RecordValueField("bool")),
      new StringDeserializer
    ).deserialize(
      "randomTopic",
      """
        |{
        |  "number": 1,
        |  "string": "hello",
        |  "bool": true
        |}""".stripMargin.getBytes(StandardCharsets.UTF_8)
    )

    val expected = Map(
      new RecordValueField("number") -> Seq(1),
      new RecordValueField("string") -> Seq("hello"),
      new RecordValueField("bool") -> Seq(true)
    )
    assert(row === expected)
  }

  test("must convert complex json type to its string representation") {
    val row = new JsonDeserializer(
      Seq(new RecordValueField("number"), new RecordValueField("record")),
      new StringDeserializer
    ).deserialize(
      "randomTopic",
      """
        |{
        |  "number": 1,
        |  "record": {
        |   "field1": "value1",
        |   "field2": 23
        |   }
        |}""".stripMargin.getBytes(StandardCharsets.UTF_8)
    )

    val expected = Map(
      new RecordValueField("number") -> Seq(1),
      new RecordValueField("record") -> Seq("""{"field1":"value1","field2":23}""")
    )
    assert(row === expected)
  }

  test("must only use fields provided to deserializer in the right order") {
    val row = new JsonDeserializer(
      Seq(new RecordValueField("record"), new RecordValueField("number")),
      new StringDeserializer
    ).deserialize(
      "randomTopic",
      """
        |{
        |  "number": 1,
        |  "fieldToIgnore": {
        |    "fieldA": 124,
        |    "fieldB": [true]
        |  },
        |  "record": {
        |   "field1": "value1",
        |   "field2": 23
        |   }
        |}""".stripMargin.getBytes(StandardCharsets.UTF_8)
    )

    val expected = Map(
      new RecordValueField("number") -> Seq(1),
      new RecordValueField("record") -> Seq("""{"field1":"value1","field2":23}""")
    )
    assert(row === expected)
  }

  test("must provide null values for fields not present") {
    val row = new JsonDeserializer(
      Seq(new RecordValueField("number"), new RecordValueField("always_null_field")),
      new StringDeserializer
    ).deserialize(
      "randomTopic",
      """
        |{
        |  "number": 1
        |}""".stripMargin.getBytes(StandardCharsets.UTF_8)
    )

    val expected = Map(
      new RecordValueField("number") -> Seq(1),
      new RecordValueField("always_null_field") -> Seq(null)
    )
    assert(row === expected)
  }

  test("must fail when all fields are referenced") {
    intercept[KafkaConnectorException] {
      new JsonDeserializer(Seq(RecordValueFields), new StringDeserializer)
        .deserialize(
          "randomTopic",
          """
            |{
            |  "number": 1
            |}
            |""".stripMargin.getBytes(StandardCharsets.UTF_8)
        )
    }
  }

  test("must produce full json when the whole value is referenced") {
    val sourceRecord =
      """
        |{
        |  "number": 1,
        |  "record": {
        |   "field1": "value1",
        |   "field2": 23
        |   }
        |}""".stripMargin

    val row = new JsonDeserializer(
      Seq(RecordValue),
      new StringDeserializer
    ).deserialize(
      "randomTopic",
      sourceRecord.getBytes(StandardCharsets.UTF_8)
    )

    assert(row.size === 1)
    assert(row.contains(RecordValue))
    val values = row(RecordValue)
    assert(values.size === 1)
    val json = values.headOption.map(_.asInstanceOf[String]).getOrElse("")
    assert(JsonMapper.fromJson[JsonNode](json) === JsonMapper.fromJson[JsonNode](sourceRecord))
  }
}

File: src/test/java/com/exasol/cloudetl/kafka/deserialization/JsonDeserializerTest.java
package com.exasol.cloudetl.kafka.deserialization;

import static org.junit.jupiter.api.Assertions.*;

import java.nio.charset.StandardCharsets;
import java.util.*;

import org.apache.kafka.common.serialization.StringDeserializer;
import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.kafka.JsonArgumentMatcher;
import com.exasol.cloudetl.kafka.KafkaConnectorException;

class JsonDeserializerTest {
    @Test
    void deserializesJsonRecordWithPrimitives() {
        final var row = deserialize(List.of(new RecordValueField("number"), new RecordValueField("string"),
                new RecordValueField("bool")), "{\"number\":1,\"string\":\"hello\",\"bool\":true}");

        assertEquals(Map.of(new RecordValueField("number"), List.of(1),
                new RecordValueField("string"), List.of("hello"),
                new RecordValueField("bool"), List.of(true)), row);
    }

    @Test
    void convertsComplexJsonTypeToStringRepresentation() {
        final var row = deserialize(List.of(new RecordValueField("number"), new RecordValueField("record")),
                "{\"number\":1,\"record\":{\"field1\":\"value1\",\"field2\":23}}");

        assertEquals(Map.of(new RecordValueField("number"), List.of(1),
                new RecordValueField("record"), List.of("{\"field1\":\"value1\",\"field2\":23}")), row);
    }

    @Test
    void onlyUsesFieldsProvidedToDeserializerInTheRightOrder() {
        final var row = deserialize(List.of(new RecordValueField("record"), new RecordValueField("number")),
                "{\"number\":1,\"fieldToIgnore\":{\"fieldA\":124,\"fieldB\":[true]},"
                        + "\"record\":{\"field1\":\"value1\",\"field2\":23}}");

        assertEquals(Map.of(new RecordValueField("number"), List.of(1),
                new RecordValueField("record"), List.of("{\"field1\":\"value1\",\"field2\":23}")), row);
    }

    @Test
    void providesNullValuesForFieldsNotPresent() {
        final var row = deserialize(List.of(new RecordValueField("number"), new RecordValueField("always_null_field")),
                "{\"number\":1}");

        final Map<FieldSpecification, List<Object>> expected = new LinkedHashMap<>();
        expected.put(new RecordValueField("number"), List.of(1));
        expected.put(new RecordValueField("always_null_field"), Collections.singletonList(null));
        assertEquals(expected, row);
    }

    @Test
    void failsWhenAllFieldsAreReferenced() {
        final List<FieldSpecification> fields = List.of(RecordValueFields.INSTANCE);
        assertThrows(KafkaConnectorException.class,
                () -> deserialize(fields, "{\"number\":1}"));
    }

    @Test
    void producesFullJsonWhenWholeValueIsReferenced() {
        final String sourceRecord = "{\"number\":1,\"record\":{\"field1\":\"value1\",\"field2\":23}}";

        final var row = deserialize(List.of(RecordValue.INSTANCE), sourceRecord);

        final var javaRow = row;
        final var values = javaRow.get(RecordValue.INSTANCE);
        assertAll(() -> assertEquals(1, javaRow.size()),
                () -> assertTrue(javaRow.containsKey(RecordValue.INSTANCE)),
                () -> assertEquals(1, values.size()),
                () -> assertEquals(JsonArgumentMatcher.readJson(sourceRecord),
                        JsonArgumentMatcher.readJson((String) values.get(0))));
    }

    private Map<FieldSpecification, List<Object>> deserialize(
            final List<FieldSpecification> fields, final String json) {
        return new JsonDeserializer(fields, new StringDeserializer()).deserialize("randomTopic",
                json.getBytes(StandardCharsets.UTF_8));
    }
}
